### PR TITLE
FHAC-94 Admin GUI Connection Manager

### PR DESCRIPTION
**Pull request by jsmith2**
Displays connection information and can ping services.

Note:
Display of wsdl is configurable.
Changes are made to both database and property files.

### CODE REVIEW

File: Product/Production/Adapters/General/CONNECTAdminGUI/src/main/java/gov/hhs/fha/nhinc/admingui/application/EndpointManager.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import java.util.Collection;
+import java.util.Date;
+
+/**
+ *
+ * @author jassmit
+ */
+public interface EndpointManager {
+    
+    public void addOrUpdateEndPoint(String url, Date timestamp, boolean pingResult);
```
Comment (christophermay07) [inline Product/Production/Adapters/General/CONNECTAdminGUI/src/main/java/gov/hhs/fha/nhinc/admingui/application/EndpointManager.java]: Camel case this to match the interface name (lowercase p)

File: Product/Production/Adapters/General/CONNECTAdminGUI/src/main/java/gov/hhs/fha/nhinc/admingui/application/EndpointManagerCache.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    private static class EndpointManagerCacheHolder {
+
+        private static final EndpointManagerCache INSTANCE = new EndpointManagerCache();
+    }
+    
+    public class EndpointCacheInfo {
+        
+        private String url;
+        private Date timestamp;
+        private boolean pingResult;
```
Comment (christophermay07) [inline Product/Production/Adapters/General/CONNECTAdminGUI/src/main/java/gov/hhs/fha/nhinc/admingui/application/EndpointManagerCache.java]: Possibly consider renaming to something along the lines of "successfulPing" or "pingable" - it gives a better indication via name of what "true" and "false" for this variable mean, and also makes the getter's name more meaningful as well (isSuccessfulPing, isPingable)

File: Product/Production/Adapters/General/CONNECTAdminGUI/src/main/java/gov/hhs/fha/nhinc/admingui/managed/ConnectionManagerBean.java [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+@ManagedBean(name = "connectionManagerBean")
+@ViewScoped
+public class ConnectionManagerBean {
+
+    private static final Logger LOG = Logger.getLogger(ConnectionManagerBean.class);
+    
+    private final HashMap<String, BusinessEntity> externalEntities = new HashMap<String, BusinessEntity>();
+    private final List<String> entityNames = new ArrayList<String>();
+    
+    //TODO Display local entity endpoints (internal and external).
```
Comment (christophermay07) [inline Product/Production/Adapters/General/CONNECTAdminGUI/src/main/java/gov/hhs/fha/nhinc/admingui/managed/ConnectionManagerBean.java:68]: Just double checking, this TODO is out of scope for 4.4.1?

Comment (jsmith2) [inline Product/Production/Adapters/General/CONNECTAdminGUI/src/main/java/gov/hhs/fha/nhinc/admingui/managed/ConnectionManagerBean.java:68]: It is out of scope as discussed tomorrow.  We’re only displaying the external entities endpoints for pinging.  Local endpoints (internal and external) will be covered in a future release, possibly 4.5.

File: Product/Production/Adapters/General/CONNECTAdminGUI/src/main/java/gov/hhs/fha/nhinc/admingui/managed/ConnectionManagerBean.java [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+    private BusinessEntity selectedEntity;
+    private String selectedEntityName;
+    
+    private List<ConnectionEndpoint> endpoints;
+    private ConnectionEndpoint selectedEndpoint;
+    
+    private final PingService pingService = new PingServiceImpl();
+    
+    private static final String HCID_TMODEL_KEY = "uddi:nhin:nhie:homecommunityid";
+    private static final String REGION_TMODEL_KEY = "uddi:uddi.org:ubr:categorization:iso3166";
```
Comment (christophermay07) [inline Product/Production/Adapters/General/CONNECTAdminGUI/src/main/java/gov/hhs/fha/nhinc/admingui/managed/ConnectionManagerBean.java]: These two UDDI strings are duplicated (privately) in CONNECTCoreLib's ConnectionManagerCacheHelper.  These should be NhincConstants -- if not as part of this PR, then we should create a ticket targeted for 4.5.0.

File: Product/Production/Adapters/General/CONNECTAdminGUI/src/main/java/gov/hhs/fha/nhinc/admingui/managed/ConnectionManagerBean.java [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+    public void init() {
+        refresh();
+    }
+    
+    public List<String> getEntityNames(){
+        return entityNames;
+    }
+    
+    public String getSelectedEntityName() {
+        String name = "--";
```
Comment (christophermay07) [inline Product/Production/Adapters/General/CONNECTAdminGUI/src/main/java/gov/hhs/fha/nhinc/admingui/managed/ConnectionManagerBean.java]: All standalone instances of "--" should be pulled out into a constant (e.g., "NULL_DISPLAY") so that we have the flexibility to change the display value later.

File: Product/Production/Adapters/General/CONNECTAdminGUI/src/main/java/gov/hhs/fha/nhinc/admingui/managed/ConnectionManagerBean.java [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+           } 
+        }
+        return hcid;
+    }
+    
+    public void ping(){
+        if(selectedEndpoint != null) {
+            boolean status = pingService.ping(selectedEndpoint.getServiceUrl());
+            Date timestamp = new Date();
+            EndpointManagerCache.getInstance().addOrUpdateEndPoint(selectedEndpoint.getServiceUrl(), timestamp, status);
```
Comment (christophermay07) [inline Product/Production/Adapters/General/CONNECTAdminGUI/src/main/java/gov/hhs/fha/nhinc/admingui/managed/ConnectionManagerBean.java]: Can we just pass "new Date()" directly in, instead of using an intermediate variable?

File: Product/Production/Adapters/General/CONNECTAdminGUI/src/main/java/gov/hhs/fha/nhinc/admingui/managed/ConnectionManagerBean.java [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+    private String getSpecVersion(CategoryBag categoryBag) {
+        if (categoryBag != null && categoryBag.getKeyedReference() != null
+                && !categoryBag.getKeyedReference().isEmpty()) {
+            for (KeyedReference kRef : categoryBag.getKeyedReference()) {
+                if (kRef.getTModelKey() != null && kRef.getTModelKey().equals(ConnectionManagerCache.UDDI_SPEC_VERSION_KEY)) {
+                    return kRef.getKeyValue();
+                }
+            }
+        }
+        return "-";
```
Comment (christophermay07) [inline Product/Production/Adapters/General/CONNECTAdminGUI/src/main/java/gov/hhs/fha/nhinc/admingui/managed/ConnectionManagerBean.java]: Should this be "--" to match the others, or is this a special case?

File: Product/Production/Adapters/General/CONNECTAdminGUI/src/main/java/gov/hhs/fha/nhinc/admingui/application/EndpointManager.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) 2009-2014, United States Government, as represented by the Secretary of Health and Human Services.
```
Comment (christophermay07) [inline Product/Production/Adapters/General/CONNECTAdminGUI/src/main/java/gov/hhs/fha/nhinc/admingui/application/EndpointManager.java:2]: Do we need to update the copyright on new files to 2015?

Comment (jsmith2) [inline Product/Production/Adapters/General/CONNECTAdminGUI/src/main/java/gov/hhs/fha/nhinc/admingui/application/EndpointManager.java:2]: I think this is out of scope.  We should have a task to update all copyrights to 2015.

**Comment by nareshcgi** [comment]
Oracle Database scripts needs to be undated.

**Comment by nareshcgi** [comment]
Reviewed the PR and looks good.

**Comment by christophermay07** [comment]
I've reviewed this PR, endorsing for merge without further comment.

**Comment by sailajaa** [comment]
Can we have ticket created in backlog to separate service logic from bean layer for ping utility. Otherwise, I am good with PR.
